### PR TITLE
Rewrite a given MediaWiki page based on template parameters

### PR DESCRIPTION
**Pull request by mhl**
[This isn't quite ready yet; e.g. there are a few tests I'd like to add, and the prompt script hasn't been updated to use this. It's probably still worth getting comments on at this stage.]

One thing to consider is whether all this new code should be in a separate gem, since it's quite a generic way of rewriting MediaWiki pages based on template parameters, and not really specific to this application.

As an example of usage, You could do:

```ruby
wiki_page = CompareWithWikidata::WikiPage.new(
  username: ENV['WIKI_USERNAME'],
  password: ENV['WIKI_PASSWORD'],
  title:   'User:Mhl20/test template'
)

class ExampleRewriter
  def rewrite(old_content:, parameters:)
    "Preserving the old content:\n#{old_content}\n" \
    "The animal parameter was: '#{parameters[:animal]}'"
  end
end

wiki_page.rewrite_and_put(ExampleRewriter.new)
```

This has been run twice on: https://www.wikidata.org/wiki/User:Mhl20/test_template for example.

### CODE REVIEW

**Comment by tmtmtmtm** [comment]
> One thing to consider is whether all this new code should be in a separate gem, since it's quite a generic way of rewriting MediaWiki pages based on template parameters, and not really specific to this application.

I would be very much in favour of separate this out like that.

File: lib/compare_with_wikidata/wiki_api.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+require 'mediawiki_api'
+
+module CompareWithWikidata
+  class WikiPage
```
Comment (mhl) [inline lib/compare_with_wikidata/wiki_api.rb]: Sure - thanks for the reference.

File: lib/compare_with_wikidata/wiki_api.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      @wikitext ||= client.get_wikitext(title).body
+    end
+
+    private
+
+    attr_accessor :username, :password, :title, :client_class
+
+    def client
+      @client ||= client_class.new('https://www.wikidata.org/w/api.php').tap do |c|
+        c.log_in(ENV['WIKI_USERNAME'], ENV['WIKI_PASSWORD'])
```
Comment (mhl) [inline lib/compare_with_wikidata/wiki_api.rb]: Oops, yes - indeed

**Comment by chrismytton** [comment]
This code has now been moved out to the [everypolitician/mediawiki-page-replaceable_content](https://github.com/everypolitician/mediawiki-page-replaceable_content) library.

**Pull request closed without merging**